run_freebsd.go

```diff
@@ -0,0 +1,572 @@
+	// will be cleaned up
+	nonCleanablePrefixes = []string{
+		"/etc", "/dev",
+	}
+)
+
+func procctl(idtype int, id int, cmd int, arg *byte) error {
+	_, _, e1 := unix.Syscall6(
+		unix.SYS_PROCCTL, uintptr(idtype), uintptr(id),
+		uintptr(cmd), uintptr(unsafe.Pointer(arg)), 0, 0)
+	if e1 != 0 {
+		return unix.Errno(e1)
+	}
+	return nil
+}
+
+func setChildProcess() error {
+	if err := procctl(P_PID, unix.Getpid(), PROC_REAP_ACQUIRE, nil); err != nil {
+		fmt.Fprintf(os.Stderr, "procctl(PROC_REAP_ACQUIRE): %v\n", err)
+		return err
+	}
+	return nil
+}
+
+func (b *Builder) Run(command []string, options RunOptions) error {
+	p, err := ioutil.TempDir("", Package)
+	if err != nil {
+		return errors.Wrapf(err, "run: error creating temporary directory under %q", os.TempDir())
+	}
+	// On some hosts like AH, /tmp is a symlink and we need an
+	// absolute path.
+	path, err := filepath.EvalSymlinks(p)
+	if err != nil {
+		return errors.Wrapf(err, "run: error evaluating %q for symbolic links", p)
+	}
+	logrus.Debugf("using %q to hold bundle data", path)
+	defer func() {
+		if err2 := os.RemoveAll(path); err2 != nil {
+			logrus.Errorf("error removing %q: %v", path, err2)
+		}
+	}()
+
+	gp, err := generate.New("freebsd")
+	if err != nil {
+		return errors.Wrapf(err, "error generating new 'freebsd' runtime spec")
+	}
+	g := &gp
+
+	isolation := options.Isolation
+	if isolation == IsolationDefault {
+		isolation = b.Isolation
+		if isolation == IsolationDefault {
+			isolation = IsolationOCI
+		}
+	}
+	if err := checkAndOverrideIsolationOptions(isolation, &options); err != nil {
+		return err
+	}
+
+	// hardwire the environment to match docker build to avoid subtle and hard-to-debug differences due to containers.conf
+	b.configureEnvironment(g, options, []string{"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"})
+
+	if b.CommonBuildOpts == nil {
+		return errors.Errorf("Invalid format on container you must recreate the container")
+	}
+
+	if err := addCommonOptsToSpec(b.CommonBuildOpts, g); err != nil {
+		return err
+	}
+
+	if options.WorkingDir != "" {
+		g.SetProcessCwd(options.WorkingDir)
+	} else if b.WorkDir() != "" {
+		g.SetProcessCwd(b.WorkDir())
+	}
+	mountPoint, err := b.Mount(b.MountLabel)
+	if err != nil {
+		return errors.Wrapf(err, "error mounting container %q", b.ContainerID)
+	}
+	defer func() {
+		if err := b.Unmount(); err != nil {
+			logrus.Errorf("error unmounting container: %v", err)
+		}
+	}()
+	g.SetRootPath(mountPoint)
+	if len(command) > 0 {
+		command = runLookupPath(g, command)
+		g.SetProcessArgs(command)
+	} else {
+		g.SetProcessArgs(nil)
+	}
+
+	setupTerminal(g, options.Terminal, options.TerminalSize)
+
+	configureNetwork, configureNetworks, err := b.configureNamespaces(g, &options)
+	if err != nil {
+		return err
+	}
+
+	containerName := Package + "-" + filepath.Base(path)
+	if configureNetwork {
+		g.AddAnnotation("org.freebsd.parentJail", containerName+"-vnet")
```

Comment:
Is there any thread for freebsd variant of runtime-spec, could we add that in comments so its easier to review.

Reply:
There isn't anything yet - Samuel Karp and I planned to setup a working group for this - hopefully starting the process some time next week. In the absence of any kind of formal FreeBSD specification, I'm using (abusing?) an annotation here for passing through network namespace information to the runtime.